src/driver-holtek/holtek8-shared.c

```diff
@@ -0,0 +1,1291 @@
+			break;
+		case HOLTEK8_API_B:
+			report.command = HOLTEK8B_CMD_READ_MACRO_DATA;
+			break;
+	}
+
+	rc = holtek8_set_feature_report(device, &report);
+	if (rc < 0)
+		return rc;
+
+	rc = holtek8_read_chunked(device, (uint8_t*) &macro_data, HOLTEK8_MACRO_DATA_SIZE, NULL);
+	if (rc < 0)
+		return rc;
+
+	while (events_i < HOLTEK8_MAX_MACRO_EVENTS) {
+		struct holtek8_macro_event *ev = &macro_data.event[data_i];
+		uint8_t *ev_data = (uint8_t*) ev;
+
+		if (ev_data[0] == 0 && ev_data[1] == 0) {
+			return 0; // macro terminator reached
+		}
+		if (ev_data[1] == HOLTEK8_MACRO_CMD_JUMP) {
+			if (single_page_macros)
+				return 0;
+
+			report.arg[0] = ev_data[0];
+			rc = holtek8_set_feature_report(device, &report);
+			if (rc < 0)
+				return rc;
+
+			rc = holtek8_read_chunked(device, (uint8_t*) &macro_data, HOLTEK8_MACRO_DATA_SIZE, NULL);
+			if (rc < 0)
+				return rc;
+
+			data_i = 0;
+		}
+		else {
+			macro_events[events_i] = *ev;
+			events_i++; data_i++;
+
+			if (data_i >= ARRAY_LENGTH(macro_data.event))
+				return 0; // end of data reached
+		}
+	}
+
+	return -EOVERFLOW;
+}
+
+/*
+ * Writes macro events to the device memory.
+ *
+ * Macros in memory are divided into pages. If the device supports jumping between
+ * pages, try to split events between multiple pages if they don't fit in one.
+ *
+ * @return An index of a first page on success or a negative errno.
+ */
+static int
+holtek8_write_macro_data(struct ratbag_device *device, struct holtek8_macro_event *macro_events)
+{
+	int rc;
+	struct holtek8_data *drv_data = device->drv_data;
+	struct holtek8_feature_report report = {0};
+	struct holtek8_macro_data macro_data = { .repeat_count = {0,1} };
+	unsigned int data_i = 0, events_i = 0;
+	unsigned int max_macro_index;
+	unsigned int events_to_write = 0;
+	unsigned int free_pages, pages_to_write;
+	uint8_t first_page = drv_data->macro_index;
+	bool single_page_macros = false;
+
+	switch (drv_data->api_version) {
+		case HOLTEK8_API_A:
+			report.command = HOLTEK8A_CMD_WRITE_MACRO_DATA;
+			max_macro_index = HOLTEK8A_MAX_MACRO_INDEX;
+			single_page_macros = true;
+			break;
+		case HOLTEK8_API_B:
+			report.command = HOLTEK8B_CMD_WRITE_MACRO_DATA;
+			max_macro_index = HOLTEK8B_MAX_MACRO_INDEX;
+			break;
+	}
+
+	report.arg[1] = HOLTEK8_MACRO_DATA_SIZE;
+
+	while (events_to_write < HOLTEK8_MAX_MACRO_EVENTS && *((uint16_t*) &macro_events[events_to_write]) != 0)
+		events_to_write += 1;
+
+	if (events_to_write == 0)
+		return -EINVAL;
+
+	free_pages = max_macro_index - drv_data->macro_index + 1;
+	pages_to_write = events_to_write / (ARRAY_LENGTH(macro_data.event) - 1);
+
+	if (events_to_write % (ARRAY_LENGTH(macro_data.event) - 1) != 0)
+		pages_to_write += 1;
+
+	if (single_page_macros && pages_to_write > 1)
+		return -ENOMEM;
+
+	if (pages_to_write > free_pages)
+		return -ENOMEM;
+
+	while (events_i < events_to_write) {
+		struct holtek8_macro_event *ev = &macro_data.event[data_i];
+		uint8_t *ev_data = (uint8_t*) ev;
```

Comment:
This looks scary. Consider making holtek8_macro_event a union with uint8_t array or maybe even another struct (considering `ev_data[1] = HOLTEK8_MACRO_CMD_JUMP;` below).

https://github.com/libratbag/libratbag/blob/a52ffcdb45a120aa3b36a5620244012d0bb0c4d8/src/driver-steelseries.c#L116-L125